File: reference/mathexpr/src/ast-types.test.ts
import { describe, test, expect } from "bun:test";
import { numberLiteral, unaryExpr, binaryExpr } from "./ast-types";

describe("ast-types", () => {
  test("numberLiteral creates a number node", () => {
    const n = numberLiteral(42);
    expect(n).toEqual({ type: "number", value: 42 });
  });

  test("unaryExpr creates a unary node", () => {
    const operand = numberLiteral(5);
    const u = unaryExpr("-", operand);
    expect(u).toEqual({
      type: "unary",
      op: "-",
      operand: { type: "number", value: 5 },
    });
  });

  test("binaryExpr creates a binary node", () => {
    const left = numberLiteral(2);
    const right = numberLiteral(3);
    const b = binaryExpr("+", left, right);
    expect(b).toEqual({
      type: "binary",
      op: "+",
      left: { type: "number", value: 2 },
      right: { type: "number", value: 3 },
    });
  });

  test("nested expressions", () => {
    // (2 + 3) * -4
    const inner = binaryExpr("+", numberLiteral(2), numberLiteral(3));
    const neg = unaryExpr("-", numberLiteral(4));
    const expr = binaryExpr("*", inner, neg);
    expect(expr.type).toBe("binary");
    expect(expr.op).toBe("*");
    expect(expr.left).toEqual({
      type: "binary",
      op: "+",
      left: { type: "number", value: 2 },
      right: { type: "number", value: 3 },
    });
    expect(expr.right).toEqual({
      type: "unary",
      op: "-",
      operand: { type: "number", value: 4 },
    });
  });
});

File: research/experiments/sweep/runs/mathexpr-full-kotlin-sonnet45/src/test/kotlin/mathexpr/AstTypesTest.kt
package mathexpr

import kotlin.test.Test
import kotlin.test.assertEquals

class AstTypesTest {
    @Test
    fun `creates number literal`() {
        val node = numberLiteral(42.0)
        assertEquals(42.0, node.value)
    }

    @Test
    fun `creates unary expression`() {
        val node = unaryExpr("-", numberLiteral(5.0))
        assertEquals("-", node.op)
        assertEquals(5.0, (node.operand as AstNode.NumberLiteral).value)
    }

    @Test
    fun `creates binary expression`() {
        val node = binaryExpr("+", numberLiteral(2.0), numberLiteral(3.0))
        assertEquals("+", node.op)
        assertEquals(2.0, (node.left as AstNode.NumberLiteral).value)
        assertEquals(3.0, (node.right as AstNode.NumberLiteral).value)
    }

    @Test
    fun `creates nested binary expression`() {
        val inner = binaryExpr("+", numberLiteral(1.0), numberLiteral(2.0))
        val outer = binaryExpr("*", inner, numberLiteral(3.0))
        assertEquals("*", outer.op)
        assertEquals("+", (outer.left as AstNode.BinaryExpr).op)
        assertEquals(3.0, (outer.right as AstNode.NumberLiteral).value)
    }
}
